test/ws.e2e-spec.ts

```diff
@@ -0,0 +1,646 @@
+
+describe('WebSocket AppController (e2e)', () => {
+  let app: INestApplication;
+  let server: ReturnType<typeof request>;
+  let http: {
+    get: ReturnType<typeof jest.fn>;
+    post: ReturnType<typeof jest.fn>;
+  };
+  let eventHandler: (events: Event[]) => void;
+  let receiptHandler: (receipt: EventStreamReply) => void;
+
+  const eventstream = {
+    connect: (
+      url: string,
+      topic: string,
+      handleEvents: (events: Event[]) => void,
+      handleReceipt: (receipt: EventStreamReply) => void,
+    ) => {
+      eventHandler = handleEvents;
+      receiptHandler = handleReceipt;
+    },
+
+    getSubscription: jest.fn(),
+  };
+
+  beforeEach(async () => {
+    http = {
+      get: jest.fn(),
+      post: jest.fn(),
+    };
+    eventstream.getSubscription.mockReset();
+
+    const moduleFixture: TestingModule = await Test.createTestingModule({
+      imports: [AppModule],
+    })
+      .overrideProvider(HttpService)
+      .useValue(http)
+      .overrideProvider(EventStreamService)
+      .useValue(eventstream)
+      .compile();
+
+    app = moduleFixture.createNestApplication();
+    app.useWebSocketAdapter(new WsAdapter(app));
+    app.useGlobalPipes(
+      new ValidationPipe({
+        whitelist: true,
+        forbidNonWhitelisted: true,
+      }),
+    );
+    await app.init();
+
+    app.get(EventStreamProxyGateway).configure('url', TOPIC);
+    app.get(TokensService).configure(BASE_URL, INSTANCE_PATH, TOPIC, PREFIX, CONTRACT_ABI, '', '');
+
+    (app.getHttpServer() as Server).listen();
+    server = request(app.getHttpServer());
+  });
+
+  afterEach(async () => {
+    await app.close();
+  });
+
+  it('Websocket: token create event', () => {
+    eventstream.getSubscription.mockReturnValueOnce(<EventStreamSubscription>{
+      name: TOPIC + ':' + CONTRACT_ADDRESS,
+    });
+
+    return server
+      .ws('/api/ws')
+      .exec(() => {
+        expect(eventHandler).toBeDefined();
+        eventHandler([mockTokenCreateEvent]);
+      })
+      .expectJson(message => {
+        expect(message.id).toBeDefined();
+        delete message.id;
+        expect(message).toEqual(mockTokenCreateWebSocketMessage);
+        return true;
+      });
+  });
+
+  it('Websocket: token create event from base subscription', () => {
+    eventstream.getSubscription.mockReturnValueOnce(<EventStreamSubscription>{
+      name: TOPIC + ':base',
+    });
+
+    return server
+      .ws('/api/ws')
+      .exec(() => {
+        expect(eventHandler).toBeDefined();
+        eventHandler([mockTokenCreateEvent]);
+      })
+      .expectJson(message => {
+        expect(message.id).toBeDefined();
+        delete message.id;
+        expect(message).toEqual(mockTokenCreateWebSocketMessage);
+        return true;
+      });
+  });
+
+  it('Websocket: token mint event', async () => {
+    eventstream.getSubscription.mockReturnValueOnce(<EventStreamSubscription>{
+      name: TOPIC + ':' + CONTRACT_ADDRESS,
+    });
+
+    http.get = jest.fn(
```

Comment:
I'm not sure this is called (it was needed for NFT URIs in ERC1155). Can probably be removed.